### ISSUE TITLE
[BUG] NullReferenceException occurs when use SKTypeface and SKPaint in multi Threading

### ISSUE THREAD

**Issue by MadhanKumarasamy** [mono/SkiaSharp#1179]
**Description**

I have try to use the SKTypeface font in the multi threading but the exception occurs. I have given the simple code sample to reproduce the issue. Kindly provide the your suggestion to resolve the issue. 

**Code**

```csharp
class Program
{
    static void Main(string[] args)
    {

        for (int i = 0; i < 50; i++)
        {
            Console.WriteLine(i);
            Thread thread = new Thread(SkiaTypefaceThreading);
            thread.Name = i.ToString();
            thread.Start();

        }

    }
    private static void SkiaTypefaceThreading()
    {
        string[] fonts = new string[] { "Arial", "Verdana", "Calibri" };
        foreach (string fontName in fonts)
        {
            FontExtension ext = new FontExtension("Arial", 12);
            Console.WriteLine(ext.Typeface.FamilyName);
        }
        FontExtension.ClearFontCache();
    }
}

public class FontExtension : SKPaint
{
    [ThreadStatic]
    private static Dictionary<string, SKTypeface> typeFaceCache;
    public static Dictionary<string, SKTypeface> TypeFaceCache
    {
        get
        {
            if (typeFaceCache == null)
                typeFaceCache = new Dictionary<string, SKTypeface>();
            return typeFaceCache;
        }
    }
    internal FontExtension(string fontName, float fontSize)
    {
        this.TextEncoding = SKTextEncoding.Utf8;
        this.IsAntialias = true;
        this.SubpixelText = true;

        if (TypeFaceCache != null && TypeFaceCache.ContainsKey(fontName + ";"))
            this.Typeface = TypeFaceCache[fontName + ";"];
        else
        {
            this.Typeface = SKTypeface.FromFamilyName(fontName, SKFontStyleWeight.Normal, SKFontStyleWidth.Normal, SKFontStyleSlant.Upright); ;

            TypeFaceCache.Add(fontName + ";", this.Typeface);
        }
        this.TextSize = fontSize;
    }
    public static void ClearFontCache()
    {

        if (typeFaceCache != null)
        {
            foreach (string key in typeFaceCache.Keys)
            {
                SKTypeface sKTypeface = typeFaceCache[key];
                if (sKTypeface != null)
                {
                    sKTypeface.Dispose();
                    sKTypeface = null;
                }
            }

            typeFaceCache.Clear();
            typeFaceCache = null;
        }
    }
}
```

**Expected Behavior**

Print the output properly.

**Actual Behavior**

Null reference exception occurs

**Basic Information**

- Version with issue:  1.59.3
- IDE:  Visual Studio 2017
- Platform Target Frameworks: .Net Core 2.1
- Target Devices:   Console Application.

**Screenshots**

![image](https://user-images.githubusercontent.com/48026813/76960846-4be0c080-6942-11ea-8acf-a276ef053a51.png)

**Comment by VijayRM** [comment]
@Redth - Any update on this?

**Comment by mattleibow** [comment]
I have pushed what I think is now a good version to the preview feed: https://aka.ms/skiasharp-eap/index.json
Check out the PR version: **1.68.2-pr.1200.20**

**Comment by mattleibow** [comment]
I had a look at the code and I think I found the issue. You have a dictionary that you are caching your fonts in - this is all good. But, when you clear the dictionary, you are disposing the objects - when they are still in the dictionary. Then, in another thread - before you remove it from the dictionary - the typeface is being used. As a result, it is setting a disposed object to the paint - which is basically null at this point.

There probably is an improvement to SkiaSharp that can be done here. Maybe we can throw if you assign a disposed object to a property. But, for a fix on your side, you can either first remove the object in the dictionary or do some locking so a read and clear don't happen together.

There are a few things to note as well. `SKTypeface.FromFamilyName(fontName)` will actually return the same object for the same typeface. This means that if you dispose it but still have a reference you actually have a reference to a disposed object. You do want to dispose typefaces that you read from a file because we need to close file handles and things, but from the matching from installed fonts can be left.

To work in your case, I would actually not use a `[ThreadStatic]` as some dictionaries might have references to the same object. INstead I would just use a plain static `ConcurrentDictionary<T, V>` and let .NET do the work. Also, I would change the clear to basically iterate over the dictionary and remove then dispose:

```csharp
public class FontExtension : SKPaint
{
    private static readonly ConcurrentDictionary<string, SKTypeface> typeFaceCache =
        new ConcurrentDictionary<string, SKTypeface>();

    public static IReadOnlyDictionary<string, SKTypeface> TypeFaceCache => typeFaceCache;

    internal FontExtension(string fontName, float fontSize)
    {
        TextEncoding = SKTextEncoding.Utf8;
        IsAntialias = true;
        SubpixelText = true;
        TextSize = fontSize;

        Typeface = typeFaceCache.GetOrAdd(
            fontName + ";",
            key => SKTypeface.FromFamilyName(fontName, SKFontStyleWeight.Normal, SKFontStyleWidth.Normal, SKFontStyleSlant.Upright));
    }

    public static void ClearFontCache()
    {
        foreach (string key in typeFaceCache.Keys)
        {
            typeFaceCache.TryRemove(key, out var typeface);
            typeface?.Dispose();
        }
    }
}
```

**Comment by mattleibow** [comment]
The particular fix for the disposing of system typefaces is fixed in #1224

**Comment by mattleibow** [comment]
Released to NuGet as version **1.68.2-preview.60**
